Realm/RLMRealm.mm

```diff
@@ -713,6 +713,58 @@ + (BOOL)performMigrationForConfiguration:(RLMRealmConfiguration *)configuration
     return success;
 }
 
+- (void)dispatchAsyncWithBlock:(void(^)(RLMRealm *))block {
+    [self dispatchAsyncWithObjects:@[] block:^(RLMRealm * _Nonnull realm, NSArray<RLMObject *> * _Nonnull) {
+        block(realm);
+    }];
+}
+
+- (void)dispatchAsyncWithObject:(RLMObject *)objectToHandOver block:(void(^)(RLMRealm *, RLMObject *))block {
+    [self dispatchAsyncWithObjects:@[objectToHandOver] block:^(RLMRealm * _Nonnull realm,
+                                                               NSArray<RLMObject *> * _Nonnull singleObjectArray) {
+        block(realm, singleObjectArray[0]);
+    }];
+}
+
+- (void)dispatchAsyncWithObjects:(NSArray<RLMObject *> *)objectsToHandOver
+                           block:(void(^)(RLMRealm *, NSArray<RLMObject *> *))block {
+    [self dispatchAsyncOnQueue:dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)
+                   withObjects:objectsToHandOver
+                         block:block];
+}
+
+- (void)dispatchAsyncOnQueue:(dispatch_queue_t)queue
+                 withObjects:(NSArray<RLMObject *> *)objectsToHandOver
+                       block:(void(^)(RLMRealm *, NSArray<RLMObject *> *))block {
+    std::vector<realm::Row> rowsToHandOver;
+    rowsToHandOver.reserve(objectsToHandOver.count);
+    NSMutableArray<NSString *> *classNames = [NSMutableArray arrayWithCapacity:objectsToHandOver.count];
+    for (RLMObject *object in objectsToHandOver) {
+        if (self != object.realm) {
+            @throw RLMException(@"Can only hand over objects from the Realm they belong to.");
+        }
+        rowsToHandOver.push_back(object->_row);
+        [classNames addObject: object.objectSchema.className];
+    }
+    std::shared_ptr<Realm::HandoverPackage> package = _realm->package_for_handover(rowsToHandOver);
+
+    RLMRealmConfiguration *config = self.configuration;
+    dispatch_async(queue, ^{
+        @autoreleasepool {
+            RLMRealm *realm = [RLMRealm realmWithConfiguration:config error:nil];
```

Comment:
We can't simply introduce a single, non-generic type since we have overloads for passing 0 objects, 1 object, or many objects. Further, these overloads are generic, leading to many different function signatures in practice. We either have to introduce a generic result type.

The motivation for the `unwrap` function is that we want to make it difficult to accidentally ignore the error. Additionally, we want to make the API as concise as possible. While I do not consider the current design particularly concise, it's better than this boilerplate:
```swift
realm.async(handingOver: [foo, bar]) { result in
    switch result {
    case let .success((realm, object)):
        try! realm.write {
            // do async writes
        }
    case let .failure(error):
        fatalError(error)
    }
}
```